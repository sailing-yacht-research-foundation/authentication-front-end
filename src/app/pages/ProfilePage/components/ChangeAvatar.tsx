import React, { useRef, useState } from 'react';
import styled from 'styled-components';
import { Auth, Storage } from 'aws-amplify';
import { getProfilePicture, getUserAttribute } from 'utils/user-utils';
import { EditFilled } from '@ant-design/icons';
import { Image, Spin, Typography } from 'antd';
import { toast } from 'react-toastify';
import Resizer from "react-image-file-resizer";

export const ChangeAvatar = (props) => {
    const { authUser } = props;

    const fileUploadRef = useRef<HTMLInputElement>(null);

    const [isUploadingProfilePicture, setIsUploadingProfilePicture] = useState<boolean>(false);

    const resizeImage = async (file) => {
        return new Promise((resolve) => Resizer.imageFileResizer(
            file, 300, 300, "PNG", 100, 0,
            (uri) => {
                resolve(uri);
            },
            'file'
        ));
    }

    const onFileChanged = async (e) => {
        e.preventDefault();
        const file = await resizeImage(e.target.files[0]);
        const avatarFileName = `${(+ new Date())}-${String(authUser.username).substring(0, 8)}-profile-picture.png`;

        setIsUploadingProfilePicture(true);

        Storage.put(avatarFileName, file, {
            contentType: "image/png",
            level: 'public',
        })
            .then(result => {
                setIsUploadingProfilePicture(false);

                Auth.currentAuthenticatedUser().then(user => {
                    Auth.updateUserAttributes(user, {
                        'picture': avatarFileName
                    }).then(response => {
                        toast.success('Upload avatar success');
                        props.cancelUpdateProfile();
                    }).catch(error => {
                        toast.error(error.message);
                    })
                }).catch(error => {
                    toast.error(error.message);
                })
            })
            .catch(err => {
                toast.error(err.message);
                setIsUploadingProfilePicture(false);
            });
    }

    const triggerChooseAvatar = () => {
        if (fileUploadRef && fileUploadRef.current) fileUploadRef.current.click();
    }

    return (
        <>
            <Spin spinning={isUploadingProfilePicture} tip="Uploading...">
                <Wrapper>
                    <AvatarHolder>
                        <Image src={getProfilePicture(authUser)} />
                    </AvatarHolder>
                    <ChangeAvatarButton>
                        <EditFilled onClick={() => triggerChooseAvatar()} size={20} />
                        <input ref={fileUploadRef} accept="image/png, image/jpeg" onChange={onFileChanged} hidden={true} type="file" />
                    </ChangeAvatarButton>
                </Wrapper>
            </Spin>
            <Typography.Title style={{ marginTop: '15px', textAlign: 'center' }} level={3}>{getUserAttribute(authUser, 'name')}</Typography.Title>
        </>
    )
}

const Wrapper = styled.div`
    position: relative;
    width: 150px;
`

const ChangeAvatarButton = styled.div`
    position: absolute;
    right: 10px;
    bottom: 20px;
    z-index: 1;
    cursor: pointer;
`

const AvatarHolder = styled.div`
    width: 150px;
    height: 150px;
    border-radius: 50%;
    overflow:hidden;
    border: 1px solid #4F61A5;
`