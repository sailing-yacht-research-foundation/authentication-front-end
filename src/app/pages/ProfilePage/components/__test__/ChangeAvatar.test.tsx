
import * as React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import { ChangeAvatar } from '../ChangeAvatar';

const shallowRenderer = createRenderer();

describe('<ChangeAvatar />', () => {
  it('should render to match the snapshot', () => {
    shallowRenderer.render(<ChangeAvatar />);
    const renderedOutput = shallowRenderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});