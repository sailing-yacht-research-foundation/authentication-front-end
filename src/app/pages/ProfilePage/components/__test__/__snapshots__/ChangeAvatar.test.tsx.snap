// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChangeAvatar /> should render to match the snapshot 1`] = `
<React.Fragment>
  <Modal
    bodyStyle={
      Object {
        "display": "flex",
        "justifyContent": "center",
        "overflow": "hidden",
      }
    }
    confirmLoading={false}
    okType="primary"
    onCancel={[Function]}
    onOk={[Function]}
    title="Change profile picture"
    visible={false}
    width={520}
  >
    <c
      backgroundColor="grey"
      borderStyle={
        Object {
          "border": "2px solid #979797",
          "borderRadius": "8px",
          "borderStyle": "dashed",
          "textAlign": "center",
        }
      }
      closeIconColor="white"
      cropColor="white"
      exportAsSquare={true}
      exportMimeType="image/png"
      exportQuality={1}
      height={295}
      label="Choose a file"
      labelStyle={
        Object {
          "color": "black",
          "cursor": "pointer",
          "display": "inline-block",
          "fontFamily": "sans-serif",
          "fontSize": "1.25em",
          "fontWeight": "700",
        }
      }
      lineWidth={4}
      mimeTypes="image/jpeg,image/png,image/gif"
      minCropRadius={30}
      mobileScaleSpeed={0.5}
      onBeforeFileLoad={[Function]}
      onClose={[Function]}
      onCrop={[Function]}
      onFileLoad={[Function]}
      onImageLoad={[Function]}
      shadingColor="grey"
      shadingOpacity={0.6}
      width={390}
    />
  </Modal>
  <Spin
    size="default"
    spinning={false}
    tip="Uploading...."
    wrapperClassName=""
  >
    <styled.div>
      <styled.div>
        <Image
          src="no-avatar.png"
          style={
            Object {
              "maxHeight": "100%",
              "maxWidth": "100%",
              "objectFit": "cover",
            }
          }
        />
      </styled.div>
      <styled.div>
        <ForwardRef(CameraFilled)
          onClick={[Function]}
          size={20}
          style={
            Object {
              "color": "#4F61A5",
              "fontSize": "25px",
            }
          }
        />
      </styled.div>
    </styled.div>
  </Spin>
</React.Fragment>
`;
